runtime/src/main/java/org/corfudb/protocols/wireprotocol/IMetadata.java

```diff
@@ -78,10 +79,30 @@ default Long getRank() {
      * @param rank The rank of this append.
      */
     default void setRank(long rank) {
+        if (rank>=0 && getRank()<0) {
+            throw new IllegalStateException("Cannot set rank when the overwrite is permitted");
```

Comment:
OK, now I think I get it.

Indeed, I believe the rank field's only purpose is to support quorum replication. Therefore, it should not have such a conflict, where some module set it to -1 and another module tries to set a positive rank.

A couple of more comments: 

I understand why you refer to a quorum operation as "overwrite", but it might be confusing, since there is no real overwrite. Rather, I think of your quorum replication implementation as transforming the stream-log from a write-once-address-space logging unit, into a ranked-address-space logging-unit (and a Paxos "acceptor"). I would suggest to comment and use names that reflect this.

 Given this, I believe that the test above should be:
     if (rank < getrank()) throw exception ..
Unless I am again missing something?

Last, to your question about rank: Again, unless I am missing something, rank is only used for quorum replication.